l3_dump.py

```diff
@@ -9,40 +9,198 @@
+        sys.exit(0)
+
+# #############################################################################
+def set_decode_loc_id():
+    """
+    Check for required LOC-decoder binary, based on environment of run
+    If L3_LOC_ENABLED=1 is set in the env, search for a LOC-decoder binary
+    named "<program_binary>_loc".
+    """
+    program_bin = sys.argv[2]
+    decode_loc_id = DECODE_LOC_ID
+    loc_decoder = LOC_DECODER
+    if LOC_ENABLED in os.environ:
+        env_var_val = os.getenv(LOC_ENABLED)
+        if env_var_val == "1":
+            loc_decoder = program_bin + "_loc"
+            if os.path.exists(loc_decoder) is False:
+                print(f"Env-var {LOC_ENABLED}=1 is set, but required "
+                      + f"LOC-decoder binary {loc_decoder} is not found.")
+                sys.exit(1)
+
+            # LOC is in-use and LOC-decoder binary was found.
+            decode_loc_id = 1
+
+    return (program_bin, decode_loc_id, loc_decoder)
+
+# #############################################################################
+def which_binary(os_uname_s:str, bin_name:str):
+    """
+    Check to see if required binary is found in $PATH. Error out otherwise.
+    """
+    if shutil.which(bin_name) is None:
+        print(f"Required {os_uname_s} binary '{bin_name}' is not found in $PATH.")
+        sys.exit(1)
+
+# #############################################################################
+def parse_rodata_start_addr(program_bin:str) -> int:
+    """
+    Parse output from `readelf` to extract start address of section '.rodata'
+
+    We are parsing the following sample hexdump output on Linux:
+
+Hex dump of section '.rodata':
+  0x00002000 01000200 00000000 2f746d70 2f6c332e ......../tmp/l3.
+  ^^^^^^^^^^
+
+    Return int for 1st hex-value; i.e., 0x00002000 .
+    """
+    with subprocess.Popen([READELF_BIN, READELF_HEXDUMP_ARG,
+                           READELF_DATA_SECTION, program_bin],
+                          stdout=subprocess.PIPE,
+                          stderr=subprocess.PIPE, text=True) as section:
+        _stdout, _stderr = section.communicate()
+
+        _rodata_offs = parse_rodata_start_offset(_stdout)
+        return _rodata_offs
+
+# #############################################################################
+def parse_rodata_start_offset(input_str:str) -> int:
+    """
+    Parse output from `readelf` to extract start offset of .rodata section.
+    Returns: Start of `.rodata` section's hex-address as int.
+    """
+    _offset = -1
+
+    # --------------------------------------------------------------------------
+    # Compile a r.e. with named capture-group to extract the 1st hex-addr.
+    # We expect output to be: 0x00002000 01000200 ...
+    # Parse leading '0x' as optional, in case `readelf` on some platforms
+    # does not spit that out.
+    # --------------------------------------------------------------------------
+    hexdump_pat = re.compile(r'(\s*)(0x)?(?P<offset_hex>[0-9a-fA-F]+)(.*)')
```

Comment:
Inline parsing done in old code using multiple `split()`s etc. is now replaced by this reliable use of r.e. with named capture-groups.